[lldb] Document the SBDebugger public interface

JDevlieghere opened the pull request:
Despite our best efforts, improved documentation of the public API remains a common request amongst its users. We have the Docstrings files in the bindings directory, but coverage and quality varies greatly by class.

I've noticed that we a bunch of Doxygen comments throughout LLDB, but relatively few in the SB API, which might be explained by the aforementioned Docstrings files. With #147617, we can leverage the Doxygen documentation to automatically generate Pydoc documentation for the Python bindings using SWIG.

Going forward, I'd like to invest in the Doxygen documentation for the public API. This PR is a first step in that direction. The Doxygen comments will benefit both C++ and Python clients. Additionally, the compiler (through the -Wdocumentation flag) can help catch quality and correctness issues.

labath reviewed (commented):
We definitely need more documentation, but I do sometimes get annoyed by it when it obscures browsing through the source code by stretching out the class definition. If that includes important information about the API, then I think it's worth it, but some of these comments tend to get very... sparse:

```
/// Gets Foo for Bar
///
/// \param [in] bar
///   Bar to get the Foo for
///
/// \return
///   Foo for bar
Foo GetFooForBar(Bar bar);
```

Now we eight lines of code for what should really be one or two. Would the documentation look too bad if we dropped the \param and \return thingies when we don't have anything meaningful to say there?

medismailben commented:
> We definitely need more documentation, but I do sometimes get annoyed by it when it obscures browsing through the source code by stretching out the class definition. If that includes important information about the API, then I think it's worth it, but some of these comments tend to get very... sparse:
> 
> ```
> /// Gets Foo for Bar
> ///
> /// \param [in] bar
> ///   Bar to get the Foo for
> ///
> /// \return
> ///   Foo for bar
> Foo GetFooForBar(Bar bar);
> ```
> 
> Now we eight lines of code for what should really be one or two. Would the documentation look too bad if we dropped the \param and \return thingies when we don't have anything meaningful to say there?

+1, may be getters and setter should be self-documented ?

medismailben commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -106,44 +124,122 @@ class LLDB_API SBDebugger {
                                           bool &is_debugger_specific);
 #endif
 
+  /// Get structured progress data from an event.
```
nit:
```suggestion
  /// Get progress structured data from an event.
```

medismailben commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -106,44 +124,122 @@ class LLDB_API SBDebugger {
+  ///
+  /// This function should be called before any other LLDB functions. It
+  /// initializes all required subsystems for proper LLDB functionality.
   static void Initialize();
```
Should this be marked as deprecated since we should favor the overload with error handling ?

JDevlieghere replied:
We could, but I don't think anything in the initialization is fallible anymore now that we don't initialize a VFS there anymore. Which reminds me, `GetReproducerPath` definitely needs to be marked as depreciated, but that's orthogonal to this PR.

medismailben commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -153,70 +249,221 @@ class LLDB_API SBDebugger {
+  ///   True to skip loading LLDB init files, false to load them.
   void SkipLLDBInitFiles(bool b);
 
+  /// Set whether to skip loading application-specific .lldbinit files.
```
Is this one specifically for Xcode ? I don't think the documentation makes it clear if that's so.

JDevlieghere replied:
No, this uses `argv[0]` and works with any tool.

bulbazord commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -292,200 +594,367 @@ class LLDB_API SBDebugger {
+  /// Clear collected statistics for targets belonging to this debugger.
+  ///
+  /// This includes clearing symbol table and debug info parsing/index time for
+  /// all modules, breakpoint resolve time and target statistics.
```
nit: Oxford comma? 😄

labath reviewed (approved):
I don't want to hold this up. Made some suggestions on how I would condense the descriptions. I stopped after I realized I'd be rewriting most of these. I also haven't checked how would the formatted version of this look like. I'll leave it up to you whether you want to apply them or not.

I have a feeling that when the users ask for documentation, they're ask for more high-level things like "what is a CompileUnit" or "what's the difference between a load and a file address", and not "what does Debugger::Create do"

jimingham commented:
> I don't want to hold this up. Made some suggestions on how I would condense the descriptions. I stopped after I realized I'd be rewriting most of these. I also haven't checked how would the formatted version of this look like. I'll leave it up to you whether you want to apply them or not.
> 
> I have a feeling that when the users ask for documentation, they're ask for more high-level things like "what is a CompileUnit" or "what's the difference between a load and a file address", and not "what does Debugger::Create do"

I agree, we need more conceptual documentation on the concepts the SB API's expose.  But I think that would be better done in a unified document (either part of the Python Reference or another Document) rather than individual high-level comments at the top of each class.  That way when something requires you to explain a couple of classes in concert, you don't have to try to figure out which class to put that information in.

JDevlieghere committed:
[lldb] Document the SBDebugger public interface

Despite our best efforts, improved documentation of the public API
remains a common request amongst its users. We have the Docstrings files
in the bindings directory, but coverage and quality varies greatly by
class.

I've noticed that we a bunch of Doxygen comments throughout LLDB, but
relatively few in the SB API, which might be explained by the
aforementioned Docstrings files. With #147617, we can leverage the
Doxygen documentation to automatically generate Pydoc documentation for
the Python bindings using SWIG.

Going forward, I'd like to invest in the Doxygen documentation for the
public API. This PR is a first step in that direction. The Doxygen
comments will benefit both C++ and Python clients. Additionally, the
compiler (through the -Wdocumentation flag) can help catch quality and
correctness issues.

JDevlieghere commented:
I think both are true. When using an IDE with LSP support, the doxygen comments can be quite helpful and they're easier to keep up to date compared to a separate design document. I've reduced the verbosity further and now almost all comments are a single line.
